### PR TITLE
[Java] Pull Consumers

### PR DESCRIPTION
**Pull request by MauriceVanVeen**
Resolves https://github.com/ConnectEverything/nats-by-example/issues/258

### CODE REVIEW

File: examples/jetstream/pull-consumer/java/Main.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        // on exit.
+        try (Connection nc = Nats.connect(natsURL)) {
+
+            // Access `JetStream` and `JetStreamManagement` which provide methods to create
+            // streams and consumers as well as convenience methods for publishing
+            // to streams and consuming messages from the streams.
+            JetStream js = nc.jetStream();
+            JetStreamManagement jsm = nc.jetStreamManagement();
+
+            // Declare a simple [limits-based stream](/examples/jetstream/limits-stream/java/).
```
Comment (scottf) [inline examples/jetstream/pull-consumer/java/Main.java:29]: Same comment as push - remove limits based note. I'm going to have a talk with Byron about this idea and probably go clean up other examples.

File: examples/jetstream/pull-consumer/java/Main.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            // (defaults to 5 seconds) is reached when not actively consuming messages.
+            // `name` is optional, if not provided it will be auto-generated.
+            // For this example, let's use the consumer with no options, which will
+            // be ephemeral with auto-generated name.
+            StreamContext streamContext = js.getStreamContext(streamName);
+            ConsumerContext consumerContext = streamContext.createOrUpdateConsumer(ConsumerConfiguration.builder().build());
+
+            // Messages can be _consumed_  continuously in callback using `consume`
+            // method. `consume` can be supplied with various options, but for this
+            // example we will use the default ones. WaitGroup is used as part of this
```
Comment (scottf) [inline examples/jetstream/pull-consumer/java/Main.java:54]: Did you mean CountDownLatch?

File: examples/jetstream/pull-consumer/java/Main.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                msg.ack();
+                latch.countDown();
+            };
+
+            try (MessageConsumer messageConsumer = consumerContext.consume(handler)) {
+                latch.await();
+
+                // Consume can be stopped by calling `stop` on the returned MessageConsumer.
+                // This will stop the callback from being called and stop retrieving the
+                // messages.
```
Comment (scottf) [inline examples/jetstream/pull-consumer/java/Main.java:69]: From the interface api doc

```
/**
 * Stop the MessageConsumer from asking for any more messages from the server.
 * The consumer will finish all pull request already in progress, but will not start any new ones.
 */
void stop();
```

**Review by scottf — CHANGES_REQUESTED** [review]

File: examples/jetstream/pull-consumer/java/Main.java (hunk 1 above)
Comment (scottf) [reply examples/jetstream/pull-consumer/java/Main.java:29]: I confirmed with Byron.

Comment (MauriceVanVeen) [reply examples/jetstream/pull-consumer/java/Main.java:29]: Changed the comment to be just:
```
// Declare a simple stream and populate it with a few messages.
```

File: examples/jetstream/pull-consumer/java/Main.java (hunk 2 above)
Comment (MauriceVanVeen) [reply examples/jetstream/pull-consumer/java/Main.java:54]: Yes, done.

File: examples/jetstream/pull-consumer/java/Main.java (hunk 3 above)
Comment (MauriceVanVeen) [reply examples/jetstream/pull-consumer/java/Main.java:69]: Adjusted to contain this.

**Review by scottf — APPROVED** [review]
LGTM

**Pull request merged**